src/main/java/gregtech/client/particle/GTParticleManager.java

```diff
@@ -18,165 +20,178 @@
+        }
         if (!newParticleQueue.isEmpty()) {
-            for (Tuple<IGTParticleHandler, GTParticle> handlerParticle = newParticleQueue.poll(); handlerParticle != null; handlerParticle = newParticleQueue.poll()) {
-                IGTParticleHandler handler = handlerParticle.getFirst();
-                GTParticle particle = handlerParticle.getSecond();
-                Map<IGTParticleHandler, ArrayDeque<GTParticle>> renderQueue = particle.getFXLayer() > 0 ? renderQueueFront : renderQueueBack;
-                if (!renderQueue.containsKey(handler)) {
-                    renderQueue.put(handler, new ArrayDeque<>());
-                }
-                ArrayDeque<GTParticle> arrayDeque = renderQueue.get(handler);
-                if (arrayDeque.size() > 6000) {
-                    arrayDeque.removeFirst().setExpired();
+            for (GTParticle particle : newParticleQueue) {
+                var queue = particle.shouldDisableDepth() ? depthDisabledParticles : depthEnabledParticles;
+
+                ArrayDeque<GTParticle> particles = queue.computeIfAbsent(particle.getRenderSetup(), setup -> new ArrayDeque<>());
+
+                if (particles.size() > 6000) {
+                    particles.removeFirst().setExpired();
                 }
-                arrayDeque.add(particle);
+                particles.add(particle);
             }
+            newParticleQueue.clear();
         }
     }
 
-    private void updateEffectLayer() {
-        if (!renderQueueBack.isEmpty()) {
-            updateQueue(renderQueueBack);
-        }
-        if (!renderQueueFront.isEmpty()) {
-            updateQueue(renderQueueFront);
-        }
-    }
-
-    private void updateQueue(Map<IGTParticleHandler, ArrayDeque<GTParticle>> renderQueue) {
-        Iterator<Map.Entry<IGTParticleHandler, ArrayDeque<GTParticle>>> entryIterator = renderQueue.entrySet().iterator();
+    private void updateQueue(Map<IRenderSetup, ArrayDeque<GTParticle>> renderQueue) {
+        var entryIterator = renderQueue.entrySet().iterator();
         while (entryIterator.hasNext()) {
-            Map.Entry<IGTParticleHandler, ArrayDeque<GTParticle>> entry = entryIterator.next();
-            Iterator<GTParticle> iterator = entry.getValue().iterator();
+            Map.Entry<IRenderSetup, ArrayDeque<GTParticle>> entry = entryIterator.next();
+            ArrayDeque<GTParticle> particles = entry.getValue();
+
+            Iterator<GTParticle> iterator = particles.iterator();
             while (iterator.hasNext()) {
-                Particle particle = iterator.next();
-                tickParticle(particle);
+                GTParticle particle = iterator.next();
+                try {
+                    particle.onUpdate();
+                } catch (RuntimeException exception) {
+                    GTLog.logger.error("particle update error: {}", particle.toString(), exception);
+                    particle.setExpired();
+                }
```

Comment:
Is this try-catch necessary? Why not hard crash on a particle exception?

Reply:
Because it's always better to fail gracefully then just explode like an EBF that touched by a raindrop 😎

While inserting try-catch may have some amount of overhead, it's not that significant, and proper error handling might lessen the damage done to player's gameplay experience (hard crashing your game vs log spam), while not sacrificing any diagnostic tools for developers. Since this is strictly client-side code, with no practical risk on actual game state if the system decides to fail catastrophically, I decided to do some error handling. Probably useless but still something, imo